gamemodes/terrortown/gamemode/server/sv_player_ext.lua

```diff
@@ -789,119 +795,263 @@ local function FindCorpse(ply)
+		self:SendRevivalReason(nil)
 
-		if not isfunction(check) or check(ply) then
-			local corpse = FindCorpse(ply)
+		self.OnReviveFailedCallback = nil
 
-			if needcorpse and (not IsValid(corpse) or corpse:IsOnFire()) then
-				ply:ChatPrint("You have not been revived because your body no longer exists.")
+		if not isfunction(DoCheck) or DoCheck(self) then
+			local corpse = FindCorpse(self)
 
-				timer.Remove(name) -- TODO needed?
+			if needsCorpse and (not IsValid(corpse) or corpse:IsOnFire()) then
+				OnReviveFailed(self, OnFail, "message_revival_failed_missing_body")
 
 				return
 			end
 
-			ply:SpawnForRound(true)
+			self:SetMaxHealth(100)
+			self:SetHealth(100)
 
-			if IsValid(corpse) then
-				local spawnPos = FindCorpsePosition(corpse)
-				if spawnPos then
-					ply:SetPos(spawnPos)
-					ply:SetEyeAngles(Angle(0, corpse:GetAngles().y, 0))
-				end
+			self:SpawnForRound(true)
+
+			if not spawnPos and IsValid(corpse) then
+				spawnPos = corpse:GetPos()
+				spawnEyeAngle = Angle(0, corpse:GetAngles().y, 0)
 			end
 
-			hook.Call("PlayerLoadout", GAMEMODE, ply)
+			spawnPos = spawnPos or self:GetDeathPosition()
+			spawnPos = spawn.MakeSpawnPointSafe(self, spawnPos)
+
+			if not spawnPos then
+				local spawnEntity = spawn.GetRandomPlayerSpawnEntity(self)
+
+				spawnPos = spawnEntity:GetPos()
+				spawnEyeAngle = spawnEntity:EyeAngles()
+			end
 
-			ply:SetMaxHealth(100)
+			self:SetPos(spawnPos)
+			self:SetEyeAngles(spawnEyeAngle or Angle(0, 0, 0))
 
-			local credits = CORPSE.GetCredits(corpse, 0)
+			hook.Run("PlayerLoadout", self, true)
 
-			ply:SetCredits(credits)
+			self:SetCredits(CORPSE.GetCredits(corpse, 0))
+			self:SelectWeapon("weapon_zm_improvised")
 
 			if IsValid(corpse) then
 				corpse:Remove()
 			end
 
-			DamageLog("TTT2Revive: " .. ply:Nick() .. " has been respawned.")
+			DamageLog("TTT2Revive: " .. self:Nick() .. " has been respawned.")
 
-			if isfunction(fn) then
-				fn(ply)
+			if isfunction(OnRevive) then
+				OnRevive(self)
 			end
-		elseif isfunction(onFail) then
-			ply:ChatPrint("Revive failed...")
-
-			onFail(ply)
+		else
+			OnReviveFailed(self, "message_revival_failed")
 		end
 	end)
 end
 
 ---
--- Returns if a player is currently in a revival process started by @{Player:Revive}
--- @return boolean The revival status
+-- Cancel the ongoing revival process.
+-- @param[default="message_revival_canceled"] string failMessage The fail message that should be displayed for the client
+-- @realm server
+function plymeta:CancelRevival(failMessage)
+	if not self:IsReviving() then return end
+
+	self:SetReviving(false)
+	self:SetBlockingRevival(false)
+	self:SendRevivalReason(nil)
+
+	timer.Remove("TTT2RevivePlayer" .. self:EntIndex())
+
+	OnReviveFailed(self, failMessage or "message_revival_canceled")
+end
+
+
+---
+-- Sets the revival state.
+-- @param[default=false] boolean isReviving The reviving state
+-- @internal
+-- @realm server
+function plymeta:SetReviving(isReviving)
```

Comment:
return if the state is the same as before. Otherwise you could have senseless networking